### PR TITLE
[IP-144] Don't save reason if status is not FAILED

### PR DESCRIPTION
**Pull request by balanza**
As a consequence of #157 we need to instruct `UserDataProcessingModel` to prevent a `reason` field to leak into an actual document stored in CosmosDB when `status != FAILED`.
A common scenario for this to happen is when, given a failed `UserDataProcessing`, its status is updated by just changing the actual value in the object.

In order to achieve this behaviour, we introduced a `beforeSave` hook to `CosmosdbModelVersioned`. Such hook is optional and, when defined, it processes a document just before being put into the database. Default behaviour is: no modifications.

### CODE REVIEW

**Comment by michaeldisaro** [comment]
@balanza I made a similar change using Omit. Take a look here https://github.com/pagopa/io-functions-commons/pull/161/files. Maybe the hook could be better though.

File: __integrations__/models/user_data_processing.test.ts [hunk 1]
```diff
@@ -140,4 +140,104 @@ describe("Models |> UserDataProcessing", () => {
...
+              status: UserDataProcessingStatusEnum.ABORTED
+            })
+          );
+          expect(result.reason).not.toBeDefined();
+        }
+      )
+      .run();
+
+    // update document with FAILED
+    const updates2 = {
```
Comment (AleDore) [inline __integrations__/models/user_data_processing.test.ts:197]: Should we rename this with a more meaningful name?

Comment (balanza) [reply __integrations__/models/user_data_processing.test.ts]: done in https://github.com/pagopa/io-functions-commons/pull/160/commits/2af7d800afeddd5a07e269158cbe1e0e09d024d6

**Review by AleDore — APPROVED** [review]
LGTM

**Pull request merged**